lib/src/command/upgrade.dart

```diff
@@ -44,14 +61,163 @@ class UpgradeCommand extends PubCommand {
+          dryRun: argResults['dry-run'],
+          precompile: argResults['precompile']);
+    }
 
     if (isOffline) {
       log.warning('Warning: Upgrading when offline may not update you to the '
           'latest versions of your dependencies.');
     }
   }
+
+  /// Writes the differences between [resolvedPackages] and the current
+  /// dependencies to the pubspec file.
+  Future<void> _writeToPubspec(
+      Map<PackageRange, PackageId> dependencyChanges) async {
+    ArgumentError.checkNotNull(dependencyChanges, 'dependencyChanges');
+
+    if (dependencyChanges.isEmpty) return;
+
+    final yamlEditor = YamlEditor(readTextFile(entrypoint.pubspecPath));
+    final initialDependencies = entrypoint.root.pubspec.dependencies.keys;
+    final initialDevDependencies = entrypoint.root.pubspec.devDependencies.keys;
+
+    for (final finalPackage in dependencyChanges.values) {
+      if (initialDependencies.contains(finalPackage.name)) {
+        yamlEditor.update(
+            ['dependencies', finalPackage.name], '^${finalPackage.version}');
+      } else if (initialDevDependencies.contains(finalPackage.name)) {
+        yamlEditor.update(['dev_dependencies', finalPackage.name],
+            '^${finalPackage.version}');
+      }
+    }
+
+    /// Windows line endings are already handled by [yamlEditor]
+    writeTextFile(entrypoint.pubspecPath, yamlEditor.toString());
+  }
+
+  /// Try to solve [pubspec] return [PackageId]'s in the resolution or `null`.
+  Future<Map<String, PackageId>> _tryResolve(Pubspec pubspec) async {
+    try {
+      final resolvedPackages = (await resolveVersions(
+        SolveType.UPGRADE,
+        cache,
+        Package.inMemory(pubspec),
+      ))
+          .packages;
+
+      final result = <String, PackageId>{};
+
+      for (final resolvedPackage in resolvedPackages) {
+        result[resolvedPackage.name] = resolvedPackage;
+      }
+
+      return result;
+    } on SolveFailure {
+      return <String, PackageId>{};
+    }
+  }
+}
+
+/// Outputs a summary of changes that will be made
+void _outputHuman(
+    Map<PackageRange, PackageId> dependencyChanges, bool isDryRun) {
+  ArgumentError.checkNotNull(dependencyChanges, 'dependencyChanges');
+
+  if (dependencyChanges.isEmpty) {
+    log.message('No breaking changes detected!');
+  } else {
+    if (isDryRun) {
+      log.message(
+          '${dependencyChanges.length} breaking change(s) would have been made:');
+    } else {
+      log.message(
+          '${dependencyChanges.length} breaking change(s) have been made:');
+    }
+
+    for (final change in dependencyChanges.entries) {
+      final initialPackage = change.key;
+      final finalPackage = change.value;
+
+      log.message('${initialPackage.name}: ${initialPackage.constraint} -> '
+          '^${finalPackage.version}');
+    }
+  }
+}
+
+/// Returns new pubspec with the same dependencies as [original] but with no
+/// version constraints on hosted packages.
```

Comment:
This is not what it does... you need to explain `useLatest` and give it a better name than `useLatest`.

Also the function shouldn't strip constraints.. it should only strip upper bounds. Let's rename it accordingly.